src/engine/enginemaster.h

```diff
@@ -126,68 +128,104 @@ class EngineMaster : public QObject, public AudioSource {
+        bool m_fadeout;
     };
 
     class GainCalculator {
       public:
         virtual double getGain(ChannelInfo* pChannelInfo) const = 0;
     };
-    class ConstantGainCalculator : public GainCalculator {
+    class PflGainCalculator : public GainCalculator {
       public:
         inline double getGain(ChannelInfo* pChannelInfo) const {
-            return pChannelInfo->m_pChannel->isTalkover() ? m_dTalkoverGain : m_dGain;
+            Q_UNUSED(pChannelInfo);
+            return m_dGain;
         }
         inline void setGain(double dGain) {
             m_dGain = dGain;
         }
-        inline void setTalkoverGain(double dGain) {
-            m_dTalkoverGain = dGain;
-        }
       private:
         double m_dGain;
-        double m_dTalkoverGain;
+    };
+    class TalkoverGainCalculator : public GainCalculator {
+      public:
+        inline double getGain(ChannelInfo* pChannelInfo) const {
+            Q_UNUSED(pChannelInfo);
+            return 1.0;
+        }
     };
     class OrientationVolumeGainCalculator : public GainCalculator {
       public:
         OrientationVolumeGainCalculator()
-                : m_dVolume(1.0), m_dLeftGain(1.0), m_dCenterGain(1.0), m_dRightGain(1.0),
-                  m_dTalkoverGain(1.0) { }
+                : m_dVolume(1.0),
+                  m_dLeftGain(1.0),
+                  m_dCenterGain(1.0),
+                  m_dRightGain(1.0) {
+        }
 
         inline double getGain(ChannelInfo* pChannelInfo) const {
-            if (pChannelInfo->m_pMuteControl->get() > 0.0) {
-                return 0.0;
-            }
-            if (pChannelInfo->m_pChannel->isTalkover()) {
-                return m_dTalkoverGain;
-            }
             const double channelVolume = pChannelInfo->m_pVolumeControl->get();
             const double orientationGain = EngineMaster::gainForOrientation(
-                pChannelInfo->m_pChannel->getOrientation(),
-                m_dLeftGain, m_dCenterGain, m_dRightGain);
+                    pChannelInfo->m_pChannel->getOrientation(),
+                    m_dLeftGain, m_dCenterGain, m_dRightGain);
             return m_dVolume * channelVolume * orientationGain;
         }
 
-        inline void setGains(double dVolume, double leftGain, double centerGain, double rightGain,
-                             double talkoverGain) {
+        inline void setGains(double dVolume, double leftGain,
+                double centerGain, double rightGain) {
             m_dVolume = dVolume;
             m_dLeftGain = leftGain;
             m_dCenterGain = centerGain;
             m_dRightGain = rightGain;
-            m_dTalkoverGain = talkoverGain;
         }
 
       private:
-        double m_dVolume, m_dLeftGain, m_dCenterGain, m_dRightGain, m_dTalkoverGain;
+        double m_dVolume;
+        double m_dLeftGain;
+        double m_dCenterGain;
+        double m_dRightGain;
+    };
+    template<typename T, unsigned int CAPACITY>
+    class FastVector {
```

Comment:
> Sorry to keep arguing, but I don't think there are any benefits that actually matter. 

For me it still matters. 

The performance gain might be minor, but there is one. We come from a highly optimized solution using 32 bit flags + a plain array with unrolled sampleutil_generated copy functions, that squeezes every unnecessary instruction. In one of my first version I had a plain array + a size integer. It was just natural to combine them into a class. 

If we switch to a QVarLengthArray we will have up to 5 * 32 = 160 limit checks when appending and 10 * 32 = 320 checks when accessing the vector on every callback. This does not fit to the rest of the code. 

> More code to maintain -> more bugs to fix.

I think the FastVector is already one of the best reviewed code. 

> Memory corruption crashes instead of clean assertion failures in debug builds.

Do not get the point, here. We have limit checks outside the Vector. 

> More work to remove them once we lift the 32-channel limit (which I plan to do today).

Why that? I am against introducing a dynamic channel count. This will come with CPU costs, for all users, where only exotic use-cases have a benefit.  I am sure we can find a suitable fixed maximum count.